File: adminweb/src/schema/ticket.ts
import { z } from "zod";

// Mirrors the backend's TicketStatus / TicketPriority. Parsing rather than
// trusting means a value the backend starts sending that we have not handled
// fails loudly here instead of rendering as a blank cell.
export const ticketStatusSchema = z.enum([
  "open",
  "in_progress",
  "pending_user",
  "resolved",
]);

export const ticketPrioritySchema = z.enum(["high", "normal", "low"]);

export const TICKET_STATUS_LABELS: Record<TicketStatus, string> = {
  open: "Open",
  in_progress: "In progress",
  pending_user: "Pending user",
  resolved: "Resolved",
};

export const TICKET_PRIORITY_LABELS: Record<TicketPriority, string> = {
  high: "High",
  normal: "Normal",
  low: "Low",
};

// The client's own list and wording, given 2026-09-04, in their order.
// "General Question" is title case while the rest are sentence case; that is
// how they wrote it.
export const TICKET_CATEGORY_OPTIONS = [
  { value: "account_access", label: "Account and access" },
  { value: "registration", label: "Registration" },
  { value: "help_student_group", label: "Help with a student or group" },
  { value: "help_mentor", label: "Help with a mentor" },
  { value: "technical_issue", label: "Technical issue" },
  { value: "certificates_records", label: "Certificates and records" },
  { value: "general_question", label: "General Question" },
  { value: "other", label: "Other" },
  // Never offered in the portal — message screening files its own tickets
  // under it. Listed here because this list also builds the queue's category
  // filter and the re-file control on the detail panel: a bucket support
  // cannot filter for is a bucket they cannot work through, and a ticket
  // screening filed here needs a way out.
  { value: "flagged_content", label: "Flagged content" },
];

/** The wording the requester picked from, not the database value.
 *
 *  The empty case still has an answer rather than falling through to a dash
 *  that reads like data was lost: screening tickets carry a category now, but
 *  rows written before it did are still readable. */
export function categoryLabel(category: string): string {
  if (!category) return "Not categorised";
  return (
    TICKET_CATEGORY_OPTIONS.find((option) => option.value === category)?.label ??
    category
  );
}

const personSchema = z.object({
  id: z.number(),
  name: z.string(),
});

export const ticketRowSchema = z.object({
  id: z.number(),
  ticketNumber: z.string(),
  user: z.object({
    // Null on tickets the platform raised itself, which have no requester.
    name: z.string().nullable(),
    region: z.string(),
    anonymous: z.boolean(),
  }),
  subject: z.string(),
  status: ticketStatusSchema,
  priority: ticketPrioritySchema,
  assignee: personSchema.nullable(),
  // The support clock, not the requester's: this column means "when did
  // anything last happen", internal notes included.
  supportUpdatedAt: z.string(),
  overdue: z.boolean(),
});

export const ticketQueueSchema = z.object({
  items: z.array(ticketRowSchema),
  total: z.number(),
  page: z.number(),
  limit: z.number(),
  hasMore: z.boolean(),
  // The two halves of one walk, and neither means anything alone.
  //
  // `asOf` is the moment this walk is a picture of: which tickets are in it
  // and in what order. `after` is the place the reader has reached, the sort
  // key of the last row on this page. Next sends both back and gets the rows
  // strictly after that place.
  //
  // Why a cursor and not the page number: the queue is ordered by last support
  // activity, which moves while somebody reads it. No ordering makes an offset
  // safe against that — a row that changes position either crosses the reader
  // or pushes the rows behind it across the reader. Measured before this
  // changed: an internal note on three of two hundred tickets, and those three
  // were served on no page while the footer still read "20 of 20".
  //
  // `after` is null on an empty page: there is no last row to continue from.
  asOf: z.string(),
  after: z.string().nullable(),
});

export const ticketSummarySchema = z.object({
  unassigned: z.number(),
  open: z.number(),
  pendingUser: z.number(),
  overdue: z.number(),
});

export const ticketAttachmentSchema = z.object({
  id: z.number(),
  filename: z.string(),
  mimeType: z.string(),
  size: z.number(),
});

export const ticketMessageSchema = z.object({
  id: z.number(),
  messageType: z.enum([
    "user_message",
    "support_reply",
    "internal_note",
    "system",
  ]),
  body: z.string(),
  author: personSchema.nullable(),
  createdAt: z.string(),
  attachments: z.array(ticketAttachmentSchema),
});

export const ticketDetailSchema = z.object({
  id: z.number(),
  ticketNumber: z.string(),
  subject: z.string(),
  body: z.string(),
  category: z.string(),
  status: ticketStatusSchema,
  priority: ticketPrioritySchema,
  channel: z.string(),
  region: z.string(),
  requester: z
    .object({
      id: z.number(),
      name: z.string(),
      email: z.string(),
      region: z.string(),
      registeredAt: z.string(),
    })
    .nullable(),
  assignee: personSchema.nullable(),
  createdAt: z.string(),
  updatedAt: z.string(),
  supportUpdatedAt: z.string(),
  firstResponseAt: z.string().nullable(),
  resolvedAt: z.string().nullable(),
  overdue: z.boolean(),
  messages: z.array(ticketMessageSchema),
});

export const ticketHistoryEntrySchema = z.object({
  id: z.number(),
  action: z.string(),
  actor: personSchema.nullable(),
  beforeState: z.unknown().nullable(),
  afterState: z.unknown().nullable(),
  createdAt: z.string(),
});

export const assigneeOptionSchema = personSchema.extend({
  /** Whether the write path will accept this person as an assignee, which is
   *  membership of the backend's `support_capable_users()`. False covers both
   *  a deactivated account and one whose support role was revoked while the
   *  account stayed active — not the same set, and reading it as only the
   *  first is the defect that offered a name `PATCH` then refused with a 400.
   *
   *  Such a person can still OWN tickets, so they belong in the filter
   *  dropdown; they must not be offered as somebody to hand work to.
   *
   *  Required, deliberately. It carried `.default(true)` and the endpoint
   *  always sends the key, so the default could only ever fire when the field
   *  went missing — and it defaulted to the permissive answer, quietly making
   *  everybody assignable, including the people this flag exists to exclude.
   *  Absent, the parse should fail loudly rather than open the dropdown up. */
  assignable: z.boolean(),
});

export const supportAgentSchema = personSchema.extend({
  email: z.string(),
  // What revoking would strand. Resolved tickets are not counted: nobody has
  // to pick those up again.
  openTickets: z.number(),
  /** The account's own status, straight from the users table.
   *
   *  Queue access and account status are kept apart on purpose, so switching
   *  an agent off leaves their row on the roster exactly where it was. The
   *  cost is that a name on this list may be an account nobody can sign into,
   *  and the endpoint sends this field so the screen can say which.
   *
   *  A plain string and not an enum, unlike status and priority above. Those
   *  two are the ticket module's own and a new value there is a defect worth
   *  failing loudly over. This one belongs to the users app, which has five
   *  values today and may gain a sixth for reasons that have nothing to do
   *  with support, and a parse failure would take the whole roster off the
   *  screen over a word this page only reads. accountStatusNote below decides
   *  what to make of it and has an answer for a value it does not know. */
  accountStatus: z.string(),
});

// Cannot sign in, so cannot work the queue. Mirrors
// users.User.INACTIVE_LOGIN_STATUSES, which deliberately leaves out "invited"
// and "pending": those are is_active=False as well, and they can still sign in
// with a usable password. Reading this off is_active instead is how an invited
// colleague gets marked as switched off.
const SIGN_IN_BLOCKED: Record<string, string> = {
  suspended: "Suspended",
  deactivated: "Deactivated",
};

/** What to say beside a roster name whose account cannot sign in, or null.
 *
 *  Null for every other status, this being a mark and not a status column: an
 *  account that can sign in is the ordinary case and needs no comment. */
export function accountStatusNote(accountStatus: string): string | null {
  return SIGN_IN_BLOCKED[accountStatus] ?? null;
}

export const ticketAuditRowSchema = z.object({
  id: z.number(),
  ticketId: z.number(),
  action: z.string(),
  actor: personSchema.nullable(),
  // Free-form JSON. For a deletion it carries the ticket number and subject,
  // which is the only way to name a ticket that no queryset returns any more.
  beforeState: z.record(z.string(), z.unknown()).nullable(),
  afterState: z.record(z.string(), z.unknown()).nullable(),
  createdAt: z.string(),
});

// Every value the ticket module actually writes to AuditLog.action. Only three
// of these are in the model's own ActionChoices — log_audit_event does not
// validate against it — so this list comes from the code that writes, not the
// enum. It also builds the audit page's action filter, which is why it is a
// list of pairs and not a map.
export const TICKET_AUDIT_ACTIONS = [
  { value: "create", label: "Created" },
  { value: "assign", label: "Assigned" },
  { value: "status", label: "Status changed" },
  { value: "priority", label: "Priority changed" },
  { value: "category", label: "Category changed" },
  { value: "resolve", label: "Resolved" },
  { value: "reopen", label: "Reopened" },
  { value: "delete", label: "Deleted" },
];

/** An audit action as words.
 *
 *  Shared by the audit page's Action column and the History list on the
 *  detail panel, which show the same AuditLog rows through two endpoints.
 *  The panel printed the stored value, so one screen said "Status changed"
 *  and the other said "status" about the same row.
 *
 *  An unrecognised value comes back unchanged rather than as a dash: the
 *  column exists to be read, and a stored word beats nothing at all. */
export function auditActionLabel(action: string): string {
  return TICKET_AUDIT_ACTIONS.find((a) => a.value === action)?.label ?? action;
}

/** Who an audit row names, when it names nobody.
 *
 *  One wording in one place because two screens print it: the Who column on
 *  the audit page, and the History list on the detail panel. Both read the
 *  same AuditLog rows through two endpoints, and until this existed the same
 *  row read "Account removed" on one screen and "system" on the other.
 *
 *  Two different things arrive here as a missing actor, and they are told
 *  apart rather than flattened into one careful sentence. AuditLog.actor_user
 *  is SET_NULL, so a row outlives the account that made it. The screening
 *  handoff writes its rows with no actor on purpose, because no person opened
 *  those tickets, and handoff.py is the only writer in the ticket module that
 *  passes actor=None — every other row had a real person on it when it was
 *  written. So a row that is not screening's and names nobody really is an
 *  account that has since been deleted, and "Account removed" is a statement
 *  about it rather than a guess.
 *
 *  Screening's rows are the ones carrying `channel` in their snapshot, which
 *  is what tells the two apart. Calling one of those a removed account
 *  invents an account that never existed, on the child-safety rows of all
 *  things.
 *
 *  The messages on the detail panel are a different question with a different
 *  answer. Those carry TicketMessage.author, which goes null for three
 *  reasons the payload cannot tell apart, so that label stays vague on
 *  purpose. See the comment on MessageRow. */
export function auditActorName(
  actor: { name: string } | null,
  afterState: unknown,
): string {
  if (actor) return actor.name;
  const after = (afterState ?? {}) as Record<string, unknown>;
  return after.channel === "ai_screening"
    ? "Automated screening"
    : "Account removed";
}

export const ticketAuditPageSchema = z.object({
  items: z.array(ticketAuditRowSchema),
  total: z.number(),
  page: z.number(),
  limit: z.number(),
  hasMore: z.boolean(),
});

const bucketSchema = z.object({ value: z.string(), count: z.number() });

export const ticketAnalyticsSchema = z.object({
  window: z.object({
    from: z.string().nullable(),
    to: z.string().nullable(),
  }),
  demand: z.object({
    volume: z.number(),
    categoryMix: z.array(bucketSchema),
    channelMix: z.array(bucketSchema),
  }),
  flow: z.object({
    unassignedBacklog: z.number(),
    ageByStatus: z.array(
      z.object({ status: z.string(), averageSeconds: z.number().nullable() }),
    ),
    reopens: z.number(),
    handOffs: z.number(),
  }),
  service: z.object({
    firstResponseSeconds: z.number().nullable(),
    answeredCount: z.number(),
    resolutionSeconds: z.number().nullable(),
    resolvedCount: z.number(),
    overdue: z.number(),
  }),
  quality: z.object({
    // Null, not zero, for an empty window: nought percent resolved is a
    // damning number to invent.
    resolutionRate: z.number().nullable(),
    resolvedCount: z.number(),
    totalCount: z.number(),
    repeatContacts: z.number(),
    satisfaction: z.number().nullable(),
    satisfactionAvailable: z.boolean(),
  }),
  segment: z
    .object({ dimension: z.string(), buckets: z.array(bucketSchema) })
    .nullable(),
  dimensions: z.array(z.string()),
});

export const regionOptionSchema = z.object({
  value: z.string(),
  label: z.string(),
});

export const bulkAssignResultSchema = z.object({
  results: z.array(
    z.object({
      ticketId: z.number(),
      ok: z.boolean(),
      error: z.string().optional(),
    }),
  ),
});

// "Nobody owns this one" as a filter value. An empty string reads as "no
// filter" on both sides, so the bucket needs a value it can actually travel
// as. Must match services/queue.py UNASSIGNED.
export const UNASSIGNED = "__unassigned__";

// The five filter dimensions plus search, all optional. An empty string means
// "no filter", matching how the backend reads them.
export const ticketFiltersSchema = z.object({
  region: z.string().optional(),
  status: z.string().optional(),
  category: z.string().optional(),
  assignee: z.string().optional(),
  priority: z.string().optional(),
  search: z.string().optional(),
});

export type TicketAnalytics = z.infer<typeof ticketAnalyticsSchema>;
export type TicketAuditRow = z.infer<typeof ticketAuditRowSchema>;
export type SupportAgent = z.infer<typeof supportAgentSchema>;
export type TicketStatus = z.infer<typeof ticketStatusSchema>;
export type TicketPriority = z.infer<typeof ticketPrioritySchema>;
export type TicketRow = z.infer<typeof ticketRowSchema>;
export type TicketQueue = z.infer<typeof ticketQueueSchema>;
export type TicketSummary = z.infer<typeof ticketSummarySchema>;
export type TicketMessage = z.infer<typeof ticketMessageSchema>;
export type TicketDetail = z.infer<typeof ticketDetailSchema>;
export type TicketHistoryEntry = z.infer<typeof ticketHistoryEntrySchema>;
export type AssigneeOption = z.infer<typeof assigneeOptionSchema>;
export type RegionOption = z.infer<typeof regionOptionSchema>;
export type TicketFilters = z.infer<typeof ticketFiltersSchema>;

// The bucket for tickets whose requester had no country on file. It needs a
// value of its own because an empty query parameter reads as "no filter"
// everywhere else on the platform.
export const UNKNOWN_REGION = "__unknown__";

File: backend/apps/tickets/models/ticket.py
from django.conf import settings
from django.db import models
from django.utils import timezone


class TicketCategory(models.TextChoices):
    """The eight self-serve topics the client specified, plus one the platform
    fills in for itself.

    The list is the client's own, given on 2026-09-04 in answer to our
    question "is Account & Access / Programs & Groups / Certificates & Records
    the full list?". It is not: those three were our guess, and the client's
    reply was "that was just an AI byproduct haha, please set the below as
    categories". So the eight below are quoted from them, in their order.

    Two things about the wording are deliberate and should not be tidied.
    "General Question" is title case while the other seven are sentence case;
    that is how the client wrote it, and it is their vocabulary to set. And
    the two survivors from the old list keep their database values
    (``account_access``, ``certificates_records``) even though their labels
    changed from "&" to "and" — the value is an internal key and the label is
    the only part anyone reads, exactly as ``in_progress`` renders as
    "In progress".

    ``programs_groups`` is gone. It has no home in the client's list, and
    migration 0005 rewrites the rows that carried it to ``other`` rather than
    to ``general_question``: "Other" is honestly the bucket for "we no longer
    have a name for this", while "General Question" is a bucket the client
    will read as meaningful on the p52 breakdown.

    FLAGGED_CONTENT is not offered to requesters: the submission serializer
    accepts only PUBLIC_TICKET_CATEGORIES (serializers.py), which is spelled
    out separately for exactly this reason, and the portal's dropdown is built
    from the same list. It exists because message screening has to file its
    tickets under something, and every requester-facing topic is a subject a
    person chose for themselves. A child-safety flag is none of those, and
    filing it under one would corrupt the category breakdown on p52, the one
    place the client asked to see what people are asking about.
    """

    ACCOUNT_ACCESS = "account_access", "Account and access"
    REGISTRATION = "registration", "Registration"
    HELP_STUDENT_GROUP = "help_student_group", "Help with a student or group"
    HELP_MENTOR = "help_mentor", "Help with a mentor"
    TECHNICAL_ISSUE = "technical_issue", "Technical issue"
    CERTIFICATES_RECORDS = "certificates_records", "Certificates and records"
    GENERAL_QUESTION = "general_question", "General Question"
    OTHER = "other", "Other"
    FLAGGED_CONTENT = "flagged_content", "Flagged content"


class TicketStatus(models.TextChoices):
    OPEN = "open", "Open"
    IN_PROGRESS = "in_progress", "In progress"
    PENDING_USER = "pending_user", "Pending user"
    RESOLVED = "resolved", "Resolved"


class TicketPriority(models.TextChoices):
    HIGH = "high", "High"
    NORMAL = "normal", "Normal"
    LOW = "low", "Low"


# The two states in which support is not waiting on anything, so the Overdue
# clock does not run.
#
# It lives beside the choices rather than in a service because both sides of
# the module read it and they read it in opposite directions: lifecycle stops
# the clock on entering one of these and starts it on leaving, while
# queue.overdue_condition excludes them from the badge. Writing either
# direction out longhand is how the two came apart twice — first by re-arming
# on every non-pending status change, then by missing the way back out of
# "resolved".
OFF_THE_CLOCK_STATUSES = (TicketStatus.PENDING_USER, TicketStatus.RESOLVED)


class TicketChannel(models.TextChoices):
    PORTAL = "portal", "Portal"
    # Reserved, deliberately not activated in the first version: the email
    # channel (inbound creation and reply threading) is a later goal, so no
    # code path writes this value yet.
    EMAIL = "email", "Email"
    AI_SCREENING = "ai_screening", "AI screening"


class Ticket(models.Model):
    """A support enquiry. Contact -> Route -> Own -> Converse -> Resolve.

    Two update clocks, deliberately (DEC-002/D1):

    ``updated_at``        what the requester sees as "last updated". Only
                          user-visible events move it.
    ``support_updated_at`` the support queue's sort key. Every event moves it,
                          including internal notes, assignment hand-offs and
                          priority changes.

    Both are ``default=timezone.now`` rather than ``auto_now`` (DEC-016①):
    ``auto_now`` stamps on every ``save()``, which would let a priority change
    bump the requester's clock and float the ticket with nothing to show for
    it in the timeline. All writes go through
    ``apps.tickets.services.lifecycle._touch``.
    """

    ticket_number = models.CharField(max_length=16, unique=True, db_index=True)
    subject = models.CharField(max_length=255)
    # No max_length here on purpose (DEC-019 A2): the 2000-character cap is a
    # serializer concern. TextField(max_length=...) would bake the cap into
    # 0001_initial and force an AlterField the moment it is tuned.
    body = models.TextField()
    category = models.CharField(
        max_length=32,
        choices=TicketCategory.choices,
        db_index=True,
    )
    status = models.CharField(
        max_length=20,
        choices=TicketStatus.choices,
        default=TicketStatus.OPEN,
        db_index=True,
    )
    priority = models.CharField(
        max_length=10,
        choices=TicketPriority.choices,
        default=TicketPriority.NORMAL,
        db_index=True,
    )
    # Snapshot of the requester's country taken at submission (DEC-006), not a
    # live join: the queue must keep filtering correctly after a user edits
    # their profile or is deleted. Width matches the source column
    # Countries.country_name(255) so the snapshot can never truncate.
    region = models.CharField(max_length=255, blank=True, db_index=True)
    # The requester's role, frozen the same way and for the same reason: the
    # dashboard segments by user type, and a live join through
    # RoleAssignmentHistory would re-label a year of history the day somebody
    # is promoted, and lose it entirely when an account is deleted
    # (created_by is SET_NULL). Blank means we never knew — including every
    # ticket raised before this column existed.
    requester_role = models.CharField(max_length=255, blank=True, db_index=True)
    channel = models.CharField(
        max_length=20,
        choices=TicketChannel.choices,
        default=TicketChannel.PORTAL,
        db_index=True,
    )
    # PROTECT, not SET_NULL, and it is a deliberate correction.
    #
    # A ticket body is free text a minor wrote about a problem they were
    # having: "I am being bullied in my group chat", an address, a phone
    # number, a parent's email. The platform already has a rule for that kind
    # of content — chat.Messages.sender_user is PROTECT, so deleting an
    # account that has ever written a message is refused, and an admin who
    # means it passes force=True, which purges the messages first
    # (apps/admin/services/user.py). Tickets were SET_NULL, so deleting a
    # student succeeded and left every word they had written sitting in the
    # support queue under no name at all. The docstring on delete_user says
    # "Delete a user and all related data"; for tickets it was not true.
    #
    # Nullable still, because a ticket raised by message screening has no
    # requester at all. Null is "there was never a person here", not "there
    # was one and they are gone" — which is also what makes the queue's
    # `anonymous` flag honest again.
    created_by = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        null=True,
        on_delete=models.PROTECT,
        related_name="tickets_created",
    )
    assignee = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        null=True,
        on_delete=models.SET_NULL,
        db_index=True,
        related_name="tickets_assigned",
    )
    first_response_at = models.DateTimeField(null=True, blank=True)
    # When the ball last landed with support, or null when it is not with
    # support at all. This is the clock the Overdue badge is measured from.
    #
    # It exists because the client rejected a first-response-only rule. Asked
    # on 2026-09-04 whether "no first reply within 4 / 24 / 72 hours" was the
    # right test, they answered: "instead of making it just the first
    # response, include follow up responses", and spelled out the loop —
    # raised, responded to, responded to by the person who raised it, and if
    # support has not answered again within the window it is overdue. The old
    # rule read ``first_response_at IS NULL``, so one reply made a ticket
    # permanently safe however long it then sat.
    #
    # A timestamp and deliberately not a boolean. Storing "is overdue" would
    # need a scheduled job to flip it and would be wrong between runs; storing
    # when the clock started needs no job at all, because the comparison
    # against the deadline still happens at read time
    # (services/queue.overdue_condition).
    #
    # Maintained by the lifecycle service, which is the only thing that writes
    # to tickets. Set on submission, on a requester's reply, on reopen, and
    # when an agent moves a ticket back off "pending user". Cleared by a
    # support reply, by "pending user", and by resolving. Deliberately NOT
    # touched by claiming, assigning or writing an internal note: none of
    # those is an answer, so none of them should make the red number go down.
    #
    # Indexed because it is a predicate on the queue's main list query.
    awaiting_support_since = models.DateTimeField(null=True, blank=True, db_index=True)
    resolved_at = models.DateTimeField(null=True, blank=True)
    created_at = models.DateTimeField(default=timezone.now)
    updated_at = models.DateTimeField(default=timezone.now, db_index=True)
    support_updated_at = models.DateTimeField(default=timezone.now, db_index=True)
    # No db_index (DEC-019 A1): all six existing soft-delete columns on the
    # platform go without one, and the column is null for almost every row.
    deleted_at = models.DateTimeField(null=True, blank=True)

    class Meta:
        db_table = 'tickets'
        verbose_name = "Ticket"
        verbose_name_plural = "Tickets"

    def __str__(self):
        return f"{self.ticket_number} ({self.status})"
